notebooks/20_markov_decision_processes_part_2/index.md

```diff
@@ -0,0 +1,672 @@
+
+$$
+\begin{aligned}
+U(s0,a0,s1,...) & = \sum_{t=0}^{\infty} \gamma^{t} R(s_{t},a_{t},s_{t+1}) \\\\
+& \leq \sum_{t=0}^{\infty} \gamma^{t} R_{max} \\\\
+U(s) & \leq \frac{R_{max}}{1-\gamma} 
+\end{aligned}
+$$
+
+If we run for $N$ iterations, we have
+$$
+\begin{aligned}
+\Vert U_{N} - U^* \Vert & = \Vert B U_{N-1} - B U^{\star} \Vert \\\\
+& \leq \gamma \Vert U_{N-1} - U^{\star} \Vert \\\\
+& = \gamma \Vert B U_{N-2} - B U^{\star} \Vert \\\\
+& \leq \gamma^{2} \Vert U_{N-2} - U^{\star} \Vert \\\\
+& \leq \gamma^{N} \Vert U_{0} - U^{\star} \Vert \\\\
+& = \gamma^{N} \Vert U^{\star} \Vert & \forall s \text{ } U_{0}(s) = 0 \\\\
+& \leq \gamma^{N} \frac{R_{max}}{1-\gamma}
+\end{aligned}
+$$
+
+if $\gamma^{N} \frac{R_{max}}{1-\gamma} \leq \epsilon$, then $\Vert U_{N} - U^{\star} \Vert \leq \epsilon$. Thus
+$$
+N = \Big \lceil \frac{\log{(\frac{R_{max}}{\epsilon (1-\gamma)})}}{\log{(\frac{1}{\gamma})}} \Big \rceil
+$$
+
+$N$ does not depend much on the ratio $\frac{\epsilon}{R_{max}}$ because value iteration converges exponentially fast but $N$ grows rapidly as $\gamma$ becomes close to 1. We can get fast convergence if we make $\gamma$ small.
+
+<html>
+<head>
+    <meta content="width=device-width, initial-scale=1" name="viewport" />
+    <style>
+                * {
+                    margin: 0;
+                    padding: 0;
+                }
+                .center-fit {
+                        max-width: 300px;
+                        max-height: 100vh;
+                        margin: auto;
+                }
+                .imgbox {
+                    height: 400px;
+                    display: grid;
+                }
+                @media only screen and (max-width: 450px) and (min-width: 0px) {
+                    .center-fit {
+                        max-width: 100%;
+                        max-height: 100vh;
+                        margin: auto;
+                    }
+                    .imgbox {
+                        height: 40%;
+                        display: grid;
+                    }
+                }
+    </style>
+</head>
+<body>
+    <div class="imgbox">
+        <img class="center-fit" src='./images/value_iteration_numbers.png'>
+    </div>
+</body>
+</html>
+
+Figure shows how $N$ varies with $\gamma$, for different values of the ratio $c = \frac{2\epsilon}{R_{max}}$. 
+
+Time complexity of each iteration of value iteration is $O(\lvert S\rvert^{2} \lvert A\rvert)$. Thus, time complexity of value iteration is $O \Big (\lvert S\rvert^{2} \lvert A\rvert \Big \lceil \frac{\log{(\frac{R_{max}}{\epsilon (1-\gamma)})}}{\log{(\frac{1}{\gamma})}} \Big \rceil \Big )$
+
+<br/>
+
+<div id='pi'><h1>Policy Iteration</h1></div>
+---
+<br/>
+
+In the previous section, we observed that it is possible to get an optimal policy even when the utility function estimate is inaccurate.
+If one action is clearly better than all others, then the exact magnitude of the utilities on the states involved need not be precise. 
```

Comment:
"need not be" seems wrong. I suggest you check your texts on Grammarly once to avoid these mistakes.